src/SDKs/Azure.Base/data-plane/Azure.Base/Http/HttpPipelineOptions.cs

```diff
@@ -3,77 +3,138 @@
+        HttpPipelineTransport _transport;
+        List<HttpPipelinePolicy> _perCallPolicies;
+        List<HttpPipelinePolicy> _perRetryPolicies;
 
-            public HttpPipelineTransport Transport {
-                get => _transport;
-                set {
-                    if (value == null) throw new ArgumentNullException(nameof(value));
-                    _transport = value;
-                }
+        public HttpPipelineTransport Transport {
+            get => _transport;
+            set {
+                if (value == null) throw new ArgumentNullException(nameof(value));
+                _transport = value;
             }
+        }
 
-            public HttpPipelinePolicy TelemetryPolicy { get; set; } = s_default;
+        public HttpPipelinePolicy LoggingPolicy { get; set; }
 
-            public HttpPipelinePolicy LoggingPolicy { get; set; } = s_default;
+        public HttpPipelinePolicy RetryPolicy { get; set; }
 
-            public HttpPipelinePolicy RetryPolicy { get; set; } = s_default;
+        public bool DisableTelemetry { get; set; } = false;
 
-            public HttpPipelinePolicy[] PerCallPolicies = Array.Empty<HttpPipelinePolicy>();
+        public string ApplicationId { get; set; }
 
-            public HttpPipelinePolicy[] PerRetryPolicies = Array.Empty<HttpPipelinePolicy>();
+        public HttpPipelineOptions( HttpPipelineTransport transport)
+            => _transport = transport;
 
-            public string ApplicationId { get; set; }
+        public HttpPipelineOptions() : this( HttpClientTransport.Shared)
+        { }
 
-            public int PolicyCount {
-                get {
-                    int numberOfPolicies = 3 + PerCallPolicies.Length + PerRetryPolicies.Length;
-                    if (LoggingPolicy == null) numberOfPolicies--;
-                    if (TelemetryPolicy == null) numberOfPolicies--;
-                    if (RetryPolicy == null) numberOfPolicies--;
-                    return numberOfPolicies;
-                }
+        public void AddPerCallPolicy(HttpPipelinePolicy policy)
+        {
+            if (policy == null) throw new ArgumentNullException(nameof(policy));
+
+            if (_perCallPolicies == null) _perCallPolicies = new List<HttpPipelinePolicy>();
+            _perCallPolicies.Add(policy);
+        }
+
+        public void AddPerRetryPolicy(HttpPipelinePolicy policy)
+        {
+            if (policy == null) throw new ArgumentNullException(nameof(policy));
+
+            if (_perRetryPolicies == null) _perRetryPolicies = new List<HttpPipelinePolicy>();
+            _perRetryPolicies.Add(policy);
+        }
+
+        public void AddService(object service, Type type = null)
+        {
+            if (service == null) throw new ArgumentNullException(nameof(service));
+
+            if (_container == null) _container = new ServiceContainer();
+            _container.Add(service, type != null ? type : service.GetType());
+        }
```

Comment:
I find this to be a bit confusing as a programming model.  If the user uses this to feed implementations to policies defined by the client or in base, the end user has to have knowledge of the specific services that the policy will honor or use.  Also, if a user uses this to feed an implementation to a custom policy they've created, is it possible that they've unintentionally changed that service for other policies in the pipeline?

Reply:
These are limitations of DI containers in general. What do you think we should do instead?